nbformat/validator.py

```diff
@@ -299,25 +454,65 @@ def validate(
         relax_add_props=relax_add_props,
         strip_invalid_metadata=strip_invalid_metadata,
     ):
+
         raise error
 
-    if notebook_supports_cell_ids:
-        # if we support cell ids check for uniqueness when validating the whole notebook
-        seen_ids = set()
-        for cell in nbdict["cells"]:
-            cell_id = cell["id"]
-            if cell_id in seen_ids:
-                if repair_duplicate_cell_ids:
-                    # Best effort to repair if we find a duplicate id
-                    cell["id"] = generate_corpus_id()
-                    get_logger().warning(
-                        "Non-unique cell id '{}' detected. Corrected to '{}'.".format(
-                            cell_id, cell["id"]
-                        )
-                    )
-                else:
-                    raise ValidationError(f"Non-unique cell id '{cell_id}' detected.")
-            seen_ids.add(cell_id)
+
+def _try_fix_error(nbdict: Any, version: int, version_minor: int, relax_add_props: bool) -> int:
+    """
+    This function try to extract errors from the validator
+    and fix them if necessary.
+
+    """
+    validator = get_validator(version, version_minor, relax_add_props=relax_add_props)
+    if not validator:
+        raise ValidationError(f"No schema for validating v{version}.{version_minor} notebooks")
+    errors = [e for e in validator.iter_errors(nbdict)]
+
+    changes = 0
+    if len(errors) > 0:
+        if validator.name == "fastjsonschema":
+            validator = get_validator(
+                version,
+                version_minor,
+                relax_add_props=relax_add_props,
+                name="jsonschema",
+            )
+            errors = [e for e in validator.iter_errors(nbdict)]
+
+        error_tree = validator.error_tree(errors)
+        if "metadata" in error_tree:
+            for key in error_tree["metadata"]:
+                nbdict["metadata"].pop(key, None)
+                changes += 1
+
+        if "cells" in error_tree:
+            number_of_cells = len(nbdict.get("cells", 0))
+            for cell_idx in range(number_of_cells):
```

Comment:
```suggestion
            for cell, cell_error in zip(nbdict.get("cells", []), error_tree["cells"]):
```
(and then replace all the `nbdict["cells"][cell_idx]` below with `cell` and the `error_tree["cells"][cell_idx]` with `cell_error`

That said, I'm saying this because I imagine that `error_tree["cells"]` is a list. If it's a dict with no assurance of being correctly ordered, we can do:
```suggestion
            for cell_idx, cell in enumerate(nbdict.get("cells", [])):
                cell_error = error_tree["cells"][cell_idx]
```

Reply:
Yeah, all that was indented/dedented as is to be extracted into its own function. I'll mark that as refactor later. Here as well I think we want to generate a copy that is fixed instead of mutating in place.